File: ios/Classes/YoloFfiPlugin.m
#import "YoloFfiPlugin.h"
#import <Foundation/Foundation.h>

// Define a separate Stream Handler class
@interface PrintStreamHandler : NSObject <FlutterStreamHandler>
@end

// The event sink is static, managed by the stream handler.
static FlutterEventSink eventSink = nil;

@implementation YoloFfiPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterEventChannel* channel = [FlutterEventChannel
        eventChannelWithName:@"com.cia1099.yolo_ffi/logging"
             binaryMessenger:[registrar messenger]];
    
    // Use a separate stream handler instance.
    PrintStreamHandler* streamHandler = [[PrintStreamHandler alloc] init];
    [channel setStreamHandler:streamHandler];
}

/**
 * @brief This is the method that will be called from C++.
 *
 * Its signature `+ (void)printMessage:(const char *)message` is designed to be
 * directly callable from the IMP in `print.cpp`, avoiding any fragile bridging
 * that caused previous crashes.
 */
+ (void)printMessage:(const char *)message {
    if (eventSink) {
        // Inside this method, we safely convert the C string to an NSString.
        NSString *nsMessage = [NSString stringWithUTF8String:message];
        if (nsMessage) {
            // Ensure the event is sent on the main thread.
            dispatch_async(dispatch_get_main_queue(), ^{
                eventSink(nsMessage);
            });
        }
    }
}

@end

#pragma mark - Stream Handler Implementation

@implementation PrintStreamHandler

- (FlutterError* _Nullable)onListenWithArguments:(id _Nullable)arguments
                                       eventSink:(FlutterEventSink)events {
    // Set the static eventSink when Dart starts listening.
    eventSink = events;
    return nil;
}

- (FlutterError* _Nullable)onCancelWithArguments:(id _Nullable)arguments {
    // Clear the static eventSink when Dart stops listening.
    eventSink = nil;
    return nil;
}
@end

File: ios/Classes/YoloFfiPlugin.swift
import Flutter

@objc(YoloFfiPlugin)
public class YoloFfiPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    // Static variable to hold the event sink.
    private static var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterEventChannel(name: "com.cia1099.yolo_ffi/print", binaryMessenger: registrar.messenger())
        let instance = YoloFfiPlugin()
        channel.setStreamHandler(instance)
    }

    // --- FlutterStreamHandler Implementation ---

    public func onListen(withArguments _: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        YoloFfiPlugin.eventSink = events
        return nil
    }

    public func onCancel(withArguments _: Any?) -> FlutterError? {
        YoloFfiPlugin.eventSink = nil
        return nil
    }

    /**
     * This class method is called from the C++ `print_message` function.
     * The `@objc` attribute makes it visible to the Objective-C runtime.
     * The method name must match the one used in `print.cpp`.
     */
    @objc
    public static func printMessage(_ message: UnsafePointer<CChar>) {
        // The incoming 'message' might be a wrapper around a C-style char*
        // whose lifetime is not guaranteed across async calls.
        // Creating a new String from it forces an immediate deep copy,
        // ensuring the data is safely managed by Swift ARC and lives long enough
        // for the async block to execute.
        let messageCopy = String(cString: message)

        guard let sink = eventSink else {
            // Optional: Add a log here if needed for debugging why the sink is nil.
            print("YoloFfiPlugin: eventSink is nil, dropping message.")
            return
        }

        // Ensure the event is sent on the main thread.
        DispatchQueue.main.async {
            sink(messageCopy)
        }
    }
}
